File: Tours/src/main/java/com/soa/toursservice/controller/ShoppingCartController.java
package com.soa.toursservice.controller;

import com.soa.toursservice.model.ShoppingCart;
import com.soa.toursservice.service.ShoppingCartService;

import org.springframework.security.core.Authentication;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.web.bind.annotation.*;
import com.soa.toursservice.model.TourPurchaseToken;

import java.util.List;

@RestController
@RequestMapping("/api/cart")
public class ShoppingCartController {

    private final ShoppingCartService shoppingCartService;

    public ShoppingCartController(ShoppingCartService shoppingCartService) {
        this.shoppingCartService = shoppingCartService;
    }

    @PostMapping("/add/{tourId}")
    public ShoppingCart addTourToCart(@PathVariable Long tourId,
                                      Authentication authentication) {

        Jwt jwt = (Jwt) authentication.getPrincipal();

        String username = jwt.getClaim("username");
        String role = jwt.getClaim("role");

        if (!"Tourist".equals(role)) {
            throw new RuntimeException("Only tourists can use shopping cart");
        }

        return shoppingCartService.addTourToCart(username, tourId);
    }

    @GetMapping
    public ShoppingCart getCart(Authentication authentication) {

        Jwt jwt = (Jwt) authentication.getPrincipal();

        String username = jwt.getClaim("username");
        String role = jwt.getClaim("role");

        if (!"Tourist".equals(role)) {
            throw new RuntimeException("Only tourists can view shopping cart");
        }

        return shoppingCartService.getCart(username);
    }

    @DeleteMapping("/items/{itemId}")
    public ShoppingCart removeItemFromCart(@PathVariable Long itemId,
                                           Authentication authentication) {

        Jwt jwt = (Jwt) authentication.getPrincipal();

        String username = jwt.getClaim("username");
        String role = jwt.getClaim("role");

        if (!"Tourist".equals(role)) {
            throw new RuntimeException("Only tourists can remove items from shopping cart");
        }

        return shoppingCartService.removeItemFromCart(username, itemId);
    }

    @PostMapping("/checkout")
    public List<TourPurchaseToken> checkout(Authentication authentication) {

        Jwt jwt = (Jwt) authentication.getPrincipal();

        String username = jwt.getClaim("username");
        String role = jwt.getClaim("role");

        if (!"Tourist".equals(role)) {
            throw new RuntimeException("Only tourists can checkout");
        }

        return shoppingCartService.checkout(username);
    }
}

File: Purchase/Purchase.API/Controllers/ShoppingCartController.cs
using Microsoft.AspNetCore.Mvc;
using Purchase.Application.Services;
using Purchase.Domain.Entities;
using System.Security.Claims;
using Microsoft.AspNetCore.Authorization;

namespace Purchase.API.Controllers;

[ApiController]
[Route("api/cart")]
[Authorize]
public class ShoppingCartController : ControllerBase
{
    private readonly ShoppingCartService _shoppingCartService;

    public ShoppingCartController(ShoppingCartService shoppingCartService)
    {
        _shoppingCartService = shoppingCartService;
    }

    [HttpPost("add/{tourId}")]
    public async Task<ActionResult<ShoppingCart>> AddTourToCart(long tourId)
    {
        var username = GetUsername();
        var role = GetRole();

        if (role != "Tourist")
        {
            return Forbid("Only tourists can use shopping cart");
        }

        var cart = await _shoppingCartService.AddTourToCartAsync(username, tourId);

        return Ok(cart);
    }

    [HttpGet]
    public async Task<ActionResult<ShoppingCart>> GetCart()
    {
        var username = GetUsername();
        var role = GetRole();

        if (role != "Tourist")
        {
            return Forbid("Only tourists can view shopping cart");
        }

        var cart = await _shoppingCartService.GetCartAsync(username);

        return Ok(cart);
    }

    [HttpDelete("items/{itemId}")]
    public async Task<ActionResult<ShoppingCart>> RemoveItemFromCart(long itemId)
    {
        var username = GetUsername();
        var role = GetRole();

        if (role != "Tourist")
        {
            return Forbid("Only tourists can remove items from shopping cart");
        }

        var cart = await _shoppingCartService.RemoveItemFromCartAsync(username, itemId);

        return Ok(cart);
    }

    [HttpPost("checkout")]
    public async Task<ActionResult<List<TourPurchaseToken>>> Checkout()
    {
        var username = GetUsername();
        var role = GetRole();

        if (role != "Tourist")
        {
            return Forbid("Only tourists can checkout");
        }

        try
        {
            var tokens = await _shoppingCartService.CheckoutAsync(username);

            return Ok(tokens);
        }
        catch (Exception ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    private string GetUsername()
    {
        return User.FindFirst("username")?.Value
            ?? User.FindFirst(ClaimTypes.Name)?.Value
            ?? throw new Exception("Username not found in token");
    }

    private string GetRole()
    {
        return User.FindFirst("role")?.Value
            ?? User.FindFirst(ClaimTypes.Role)?.Value
            ?? throw new Exception("Role not found in token");
    }

    [HttpGet("has-purchased/{tourId}")]
    public async Task<ActionResult<bool>> HasPurchased(long tourId)
    {
        var username = GetUsername();
        var role = GetRole();

        if (role != "Tourist")
        {
            return Forbid("Only tourists can check purchased tours");
        }

        var hasPurchased = await _shoppingCartService.HasPurchasedAsync(username, tourId);

        return Ok(hasPurchased);
    }

    [HttpGet("purchased/{touristUsername}")]
    public async Task<IActionResult> GetPurchasedTours(string touristUsername)
    {
        try
        {
            var purchasedTours = await _shoppingCartService.GetPurchasedToursAsync(touristUsername);

            return Ok(purchasedTours);
        }
        catch (Exception ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }
}
